### DISCUSSION TITLE
How to conveniently plots 4 character integers as years (temporal data)

### DISCUSSION THREAD

**Discussion by joelostblom** [vega/altair#3140] [Q&A] [answered]
This is an Altair user question xD What is a good way of visualizing years stored as integers? If I naively plot them "as is", the formatting with the thousands separator "," will be unfortunate and I have to add `scale(zero=False)`:

```py
import altair as alt
from vega_datasets import data


source = data.wheat()

alt.Chart(source).mark_point().encode(
    x=alt.X("year:Q").scale(zero=False),  # could fix the formatting with `.axis(format='i')` but that + scale is a lot of typing
    y='wheat',
)
```

![image](https://github.com/altair-viz/altair/assets/4560057/d5a547c1-35f7-4ed7-b057-f1693790948b)

Another way would be to treat the years as ordinal, but that creates a wide chart with a redundant amount of axis labels:

```py
alt.Chart(source, width=600).mark_point().encode(
    x=alt.X("year:O"),
    y='wheat',
)
```

![image](https://github.com/altair-viz/altair/assets/4560057/441e3a46-75be-425e-9d48-e42f5979fea2)

I could convert them in pandas, but that is syntactically cumbersome and also stores them incorrectly in the pandas dataframe as `YYYY-MM-DD`  which implies that I have information on  the month and date although I don't:

```py
import pandas as pd
# `to_datetime doesn't work for older dates, but I think `parse_dates` in `read_csv` would
alt.Chart(source.assign(year=pd.to_datetime(source['year'], format='%Y', errors='coerce'))).mark_point().encode(
    x="year",
    y='wheat',
)
```

![image](https://github.com/altair-viz/altair/assets/4560057/4a69c2c2-b07e-49e7-91cc-41617903a8b9)

I wish I could just append `:T` and that Vega-Lite would figure out that 4 characters is a year automatically, but that doesn't happen:

```py
alt.Chart(source).mark_point().encode(
    x="year:T",
    y='wheat',
)
```

![image](https://github.com/altair-viz/altair/assets/4560057/75fc8add-6e18-43b7-8a10-f84ce7362ba0)

And the `timeUnit` transforms don't seem helpful here either as they operate on legitimate date strings and not 4-character year integers. I am hoping that there is a straightforward way to plot/format year integers as dates but I can't seem to find it myself. Any pointers would be much appreciated.

**Comment by binste** [comment]
Yeah that would be nice. The shortest I can come up with is converting it to a datetime and back with Vega expressions:

```python
alt.Chart(source).mark_point().transform_calculate(
    x_date="datetime(datum.year, 0)"
).encode(
    x="year(x_date):T",
    y="wheat",
)
```

**Reply by joelostblom** [reply]
Thank you, that's a good alternative, but still pretty complicated to e.g. explain to a novice of Altair. Maybe I'll open an issue on the Vega-Lite repo later to see if there would be room to improve this.

**Reply by mattijn** [reply, marked as answer]
Use `str` as `dtype` for the `year` column in your dataframe works as well:

```python
import altair as alt
from vega_datasets import data

source = data.wheat()
source.year = source.year.astype(str)

alt.Chart(source).mark_point().encode(
    x='year:T',
    y='wheat',
)
```

Not sure how you could know this upfront.

**Reply by mattijn** [reply]
(I was reading https://vega.github.io/vega-lite/docs/timeunit.html#example-customizing-step and thought I might try something similar here.)

**Reply by joelostblom** [reply]
Ohhh wow that's great! Thank you @mattijn ! I opened an issue to add this to the documentation. https://github.com/altair-viz/altair/issues/3162

**Reply by joelostblom** [reply]
I came a cross this recently in VegaFusion https://github.com/hex-inc/vegafusion/issues/402, so I dug up an old Vega issue and asked if it would be possible to change so that ints with a temporal dtype are interpreted as years https://github.com/vega/vega/issues/1681. If that is not possible, I think we could consider doing this on the altair side of things and recast column automatically, but it would need to be in a breaking change release. We actually have an old altair issue for this as well https://github.com/altair-viz/altair/issues/1365.
